src/mme-app/actionHandlers/s1HandoverActionHandlers.cpp

```diff
@@ -0,0 +1,440 @@
+/*
+ * Copyright 2020-present Infosys Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/******************************************************************************
+ *
+ * This file has both generated and manual code.
+ * 
+ * File template used for code generation:
+ * <TOP-DIR/scripts/SMCodeGen/templates/stateMachineTmpls/actionHandlers.cpp.tt>
+ *
+ ******************************************************************************/
+#include "actionHandlers/actionHandlers.h"
+#include "controlBlock.h"
+#include "msgType.h"
+#include "procedureStats.h"
+#include "log.h"
+#include "secUtils.h"
+#include "state.h"
+#include <string.h>
+#include <sstream>
+#include <mmeSmDefs.h>
+#include "common_proc_info.h"
+#include <ipcTypes.h>
+#include <tipcTypes.h>
+#include <msgBuffer.h>
+#include <interfaces/mmeIpcInterface.h>
+#include <event.h>
+#include <stateMachineEngine.h>
+#include <utils/mmeCommonUtils.h>
+#include <utils/mmeContextManagerUtils.h>
+#include <utils/mmeS1MsgUtils.h>
+#include <utils/mmeGtpMsgUtils.h>
+#include <contextManager/dataBlocks.h>
+
+using namespace mme;
+using namespace SM;
+using namespace cmn::utils;
+
+extern MmeIpcInterface *mmeIpcIf_g;
```

Comment:
Wondering why Singleton design pattern  is not used. Why to use extern. Does not look good.

Reply:
Singleton is usually used for classes that contain config storage or containers etc. I will create a component DB class for this and we can access components like this via the db class. Will do it as part of the current prop activity.

Reply:
Hmm...Is there any existing component class. May be pass me reference, I would like to see it once.  Since message interface is single i though singleTon is better than extern. But let me see what is component class.

Reply:
I meant something like this:

**_componentDb.h_** 
```
class ComponentIf{
public:
  virtual ~ComponentIf() {};
};

class ComponentDb{
public:
  bool registerComponent(Uint16 componentId, ComponentIf* comp);
  ComponentIf& getComponent(Uint16 componentId);
  
private:
  std::map<Uint16, ComponentIf*> componentContainer;

};

extern ComponentDb compDb;
```
**_mmeIpcInterface.h_**
```
class MmeIpcInteface : public ComponentIf
{... }

```
**_mmeIpcInterface.cpp_**
```
MmeIpcInterface::MmeIpcInterface()
{
   ...
   registerComponent(MmeIpcIf, this)
   ...
   
}
```
**_ActionHandler.cpp_**
```
#include <componentDb.h>
....
compDb.getComponent(MmeIpcIf)
```